blog/2020-11-05-HowMuchFasterIsJava15.adoc

```diff
@@ -0,0 +1,105 @@
+= How much faster is Java 15?
+:page-interpolate: true
+:awestruct-author: mtomco
+:awestruct-layout: blogPostBase
+:awestruct-tags: [production, benchmark]
+
+Java 15 was released on the 15th of September 2020 and has promised to bring with it a few performance tweaks in its G1GC
+and ParallelGC garbage collectors. What does it mean for
+https://www.optaplanner.org/[OptaPlanner]? Have the improvements made in the G1GC outperform the reigning champion since Java 8 or does the ParallelGC still
+reign supreme on most use-cases? Is ParallelGC even faster than it was the previous time? Let's put it to the test!
+
+This article is a follow up on our
+https://www.optaplanner.org/blog/2019/01/17/HowMuchFasterIsJava11.html[ Java 11]
+and our
+https://www.optaplanner.org/blog/2014/03/20/HowMuchFasterIsJava8.html[Java 8] performance comparison benchmarks.
+
+== Benchmark methodology
+
+To run the benchmark we used:
+
+* A stable machine without any other computational demanding processes running and with
+`Intel® Xeon® Silver 4116 @ 2.1 GHz (12 cores total / 24 threads)` and `128 GiB` RAM memory, running `RHEL 8 x86_64`.
+
+* Both G1 and Parallel GC for both Java versions to compare the impact of garbage collection.
+Java executed with the parameters `-Xmx1536M -server -XX:+UseG1GC` and `-Xmx1536M -server -XX:+UseParallelGC` respectively.
+
+* Both OpenJDK 11 version "11.0.6"
+OpenJDK Runtime Environment 18.9 (build 11.0.6+10-LTS)
+OpenJDK 64-Bit Server VM 18.9 (build 11.0.6+10-LTS, mixed mode)
+
+* Both OpenJDK 15 version "15.0.1"
+OpenJDK Runtime Environment 18.9 (build 15.0.1+9-LTS)
+OpenJDK 64-Bit Server VM 18.9 (build 15.0.1+9-LTS, mixed mode)
+
+* OptaPlanner `7.44.0.Final`
+
+* Solving a planning problem involves *no IO* (except a few milliseconds during startup to load the input). *A single
+CPU is completely saturated.* It constantly creates many short-lived objects, and the GC collects them afterwards.
+
+* Each run solves 11 planning problems with OptaPlanner. Each planning problem runs for 5 minutes and starts with a
+30 second JVM warm up which is discarded.
```

Comment:
So have we use the OptaPlanner Benchmarker for this?
If so, I can see people striking the results down, as that is very far from giving statistically significant, reproducible results.

Reply:
Can we just say how much warm-up time is enough? I don't think so. Let's use statistics to help us:
- how many iterations have we tried?
- between these iterations, how much do the results vary?

If we prolong the warm-up period, do the results change much?

Reply:
That's my point, though - we don't have the concept of "iterations" for the Benchmarker.
If this was done with JMH, I am wrong - but I do not see that written anywhere.

Reply:
AFAIK it was truly run using the OptaPlanner Benchmark. Thus, we have to take care of iterations ourselves. I think @michaltomco mentioned he ran the benchmarks multiple times.

Reply:
@rsynek So there is no way for me to easily reproduce this result. This is not good.

Reply:
@triceo well, JMH does not measure the score calculation speed. For this, we at least have the pre-configured ["run everything" Benchmark](https://github.com/kiegroup/optaplanner/blob/master/optaplanner-examples/src/main/resources/org/optaplanner/examples/app/benchmark/generalOptaPlannerBenchmarkConfig.xml).

Reply:
I see your point.
Still, I think that "How do I run this?" is going to be the first question on the mind of a curious reader.
And "Take this, and run it 20 different times." is not a satisfactory answer, I'm afraid.

Reply:
We can go one step further and offer a script that runs it, including the iterations. Yet, completing the results would be left to the curious reader.

Even better, we could (and should) improve optaplanner-benchmark to take care of such iterations and showing standard deviation as JMH does, so that we can tell if the warmup was sufficient or not. But that's definitely out of scope for this blogpost.

Regardless of what we do, we can't make the results 100% reproducible as they are bound to a specific testing environment.

Reply:
We agreed with @michaltomco to add a hint about the general benchmark app which was used to get the results.